### PR TITLE
Cross-contract Call Attack Prevention Standard

### PR DESCRIPTION
**Pull request by WyattMufson**
This proposal introduces a standard to protect contracts from cross-contract call attacks.

### CODE REVIEW

**Comment by EdgeDLT** [comment]
Hey Wyatt, good to see you bringing some of those ideas back over to team green 😉

From what I can understand, it seems the planned [Manifest and Permission system](https://github.com/neo-project/neo/issues/287) would resolve the same problem in a more comprehensive manner. Can you verify?

**Comment by vncoelho** [comment]
Exactly, @edgedlt.
I believe it is more a requeriment than a proposal, the ABI implementation is specifing constraints.

**Comment by WyattMufson** [comment]
@EdgeDLT ah hadn't seen that before. Is the manifest able to be updated after deploying? And if so, does it cost 100-1000 GAS like Contract.Migrate?

Seems like the manifest probably solves the issue more than this proposal.

**Comment by EdgeDLT** [comment]
Implementation is [here ](https://github.com/neo-project/neo/pull/766)if you'd like to take a look. [Apparently](https://github.com/neo-project/neo/pull/766#issuecomment-494219076), permissions will be updatable without migration.

**Comment by WyattMufson** [comment]
Yeah looks good to me. Would just be curious on the cost/difficulty to update the manifest.

**Comment by WyattMufson** [comment]
It looks like `Contract.Update` costs the same as `Contract.Create`. While the manifest is more comprehensive, it does look like it would be too expensive with the current fee structure: https://github.com/neo-project/neo/blob/master/neo/SmartContract/InteropService.NEO.cs#L31

**Comment by WyattMufson** [comment]
Also is this included in NEO 2.x or just in 3.0?

**Comment by EdgeDLT** [comment]
I expect just 3.0, 2.x is LTS now it seems.

Recommend checking against the 3.0 fee structure.

**Comment by WyattMufson** [comment]
@EdgeDLT Hm. Would this standard have usage for NEO 2.x, since 3.0 is still a year away? Any idea what the fee will be for `Contract.Update`? @erikzhang ?

**Comment by shargon** [comment]
Manifest and permission system are only for neo3

**Comment by WyattMufson** [comment]
So then this looks like a good solution for NEO 2.x, right? @EdgeDLT what are your thoughts?

@shargon is it clear yet what the gas fee will be? Doesn't it look like `Contract.Update` and `Contract.Create` will be the [same gas fee](https://github.com/neo-project/neo/blob/master/neo/SmartContract/InteropService.NEO.cs#L31)? If this is prohibitively high, like in 2.x then I think this NEP could be useful in 3.0.

If 3.0 is not prohibitively expensive to update, what would people's thoughts be? I know 2.x is not getting new features but that doesn't mean projects aren't planning on launching on NEO before 3.0.

**Comment by EdgeDLT** [comment]
I think we should move ahead on this.

When a NEO3 TestNet exists we can probably assume that most projects will focus on building for that. Until then, there are probably going to be others deploying or migrating contracts on NEO 2.x. There's not really a downside with providing a standard that can offer them some extra security.

**Comment by WyattMufson** [comment]
I agree with @EdgeDLT on this

**Comment by WyattMufson** [comment]
@erikzhang @shargon

File: nep-ccap.mediawiki [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+The <code>RequireApproved</code> method takes no arguments.
+If the results of <code>GetCallingScriptHash()</code> and <code>GetEntryScriptHash()</code> do not match, then the <code>callerHash</code> must be in the whitelist for the wallet.
+If not, this method must throw an exception.
+
+==Rationale==
+
+This NEP was designed with the usability and security of digital assets on the NEO blockchain in-mind.
+The goal is to enable a new type of asset on NEO that will be optimized for use in smart contracts while maintaining the security of its users.
+Contracts following this NEO are meant to also be following a standard for other crypto assets - such as NEP5, NEP10 and NEP11.
```
Comment (corollari) [inline nep-ccap.mediawiki]: NEO -> NEP

**Pull request closed without merging**